internal/api/provider/azure.go

```diff
@@ -54,58 +62,73 @@ func NewAzureProvider(ext conf.OAuthProviderConfiguration, scopes string) (OAuth
 			RedirectURL: ext.RedirectURI,
 			Scopes:      oauthScopes,
 		},
-		APIPath: apiPath,
+		ExpectedIssuer: expectedIssuer,
 	}, nil
 }
 
 func (g azureProvider) GetOAuthToken(code string) (*oauth2.Token, error) {
 	return g.Exchange(context.Background(), code)
 }
 
-func (g azureProvider) GetUserData(ctx context.Context, tok *oauth2.Token) (*UserProvidedData, error) {
-	var u azureUser
-	if err := makeRequest(ctx, tok, g.Config, g.APIPath+"/oidc/userinfo", &u); err != nil {
-		return nil, err
+func (g azureProvider) detectIDTokenIssuer(ctx context.Context, idToken string) (string, error) {
+	var payload struct {
+		Issuer string `json:"iss"`
 	}
 
-	var data UserProvidedData
-
-	data.Metadata = &Claims{
-		Issuer:  g.APIPath,
-		Subject: u.Sub,
-		Name:    u.Name,
+	parts := strings.Split(idToken, ".")
+	if len(parts) != 3 {
+		return "", fmt.Errorf("azure: invalid ID token")
+	}
 
-		// To be deprecated
-		FullName:   u.Name,
-		ProviderId: u.Sub,
+	payloadBytes, err := base64.RawURLEncoding.DecodeString(parts[1])
+	if err != nil {
+		return "", fmt.Errorf("azure: invalid ID token %w", err)
 	}
 
-	if u.Email != "" {
-		data.Emails = append(data.Emails, Email{
-			Email:    u.Email,
-			Verified: true,
-		})
+	if err := json.Unmarshal(payloadBytes, &payload); err != nil {
+		return "", fmt.Errorf("azure: invalid ID token %w", err)
 	}
 
-	if u.OtherMails != nil {
-		for _, mail := range u.OtherMails {
-			if mail != "" {
-				data.Emails = append(data.Emails, Email{
-					Email:    mail,
-					Verified: false,
-				})
-			}
+	return payload.Issuer, nil
+}
+
+func (g azureProvider) GetUserData(ctx context.Context, tok *oauth2.Token) (*UserProvidedData, error) {
+	idToken := tok.Extra("id_token")
+
+	fmt.Printf("AZURE ID TOKEN %v\n", idToken)
+
+	if idToken != nil {
+		issuer, err := g.detectIDTokenIssuer(ctx, idToken.(string))
+		if err != nil {
+			return nil, err
 		}
-	}
 
-	if len(data.Emails) == 0 {
-		return nil, errors.New("unable to find email with Azure provider")
-	}
+		if !IsAzureIssuer(issuer) {
+			return nil, fmt.Errorf("azure: ID token issuer not valid %q", issuer)
+		}
+
+		if g.ExpectedIssuer != "" && issuer != g.ExpectedIssuer {
+			return nil, fmt.Errorf("azure: ID token issuer %q does not match expected issuer %q", issuer, g.ExpectedIssuer)
+		}
```

Comment:
i don't think we need to check the issuer here, the oidc library already handles it for us [here](https://github.com/supabase/gotrue/blob/482cbff9a7a6a52af70544e7bdc4cd87a7ce82f8/internal/api/provider/oidc.go#L38)

Reply:
We do, as I said if you call the single tenant authorize the OIDC token expected back must be from that issuer. If you use the common authorize, there can be any number of issuers.